src/playwright/matchers/to-dispatch-events.ts

```diff
@@ -12,115 +12,148 @@ interface SerializableEvent {
+      // event listeners are in place before the action call is awaited. And the only
+      // way to ensure that is for the matcher to first await this evaluation.
+      //
+      // Before you scoff, as I did, at adding a property to the window, consider how
+      // useful this matcher is for testers. And consider whether this window nonsense
+      // will be flaky or unreliable in any way.
+      const windowWithEvents = window as typeof window & {
+        events: SerializableEvent[];
+      };
+
+      windowWithEvents.events = [];
+
+      // A Set to deduplicate events. So only one listener is added per event type.
+      const eventTypes = new Set<string>(
+        expectedEvents.map(({ type }) => type),
+      );
 
-            // The promise is resolved above when an event is dispatched. It also needs to be
-            // resolved after a timeout because some tests assert that no events were
-            // dispatched after an action.
-            //
-            // `5000` because actions can take some time to execute, especially in CI.
-          }, 5000);
-        });
+      for (const type of eventTypes) {
+        element.addEventListener(type, (event: Event) => {
+          const windowWithEvents = window as typeof window & {
+            events: SerializableEvent[];
+          };
 
-        // Wait a tick to catch duplicate events, which tests assert against.
-        return new Promise<SerializableEvent[]>((resolve) => {
-          setTimeout(() => {
-            resolve(events);
+          windowWithEvents.events.push({
+            bubbles: event.bubbles,
+            cancelable: event.cancelable,
+            composed: event.composed,
+            defaultPrevented: event.defaultPrevented,
+            timeStamp: event.timeStamp,
+            type: event.type,
           });
         });
-      });
-
-      const results = await Promise.allSettled(promises);
-
-      return results
-        .filter(
-          (result): result is PromiseFulfilledResult<SerializableEvent[]> => {
-            return result.status === 'fulfilled';
-          },
-        )
-        .flatMap(({ value }) => value);
+      }
     }, expectedEvents);
 
     await action();
 
-    const receivedEvents = await receivedEventsPromise;
+    await new Promise((resolve) => {
+      // Timeouts are obviously a source of flakiness. But actions can take some time
+      // to run, especially in CI.
+      //
+      // Try increasing this number slightly if you see a test passing locally but
+      // failing in CI. Most likely not enough time has passed for the listener to be
+      // called after the action.
+      //
+      // The magic number is the lowest one that results in no flakiness. Any test that
+      // uses this matcher will take at least as long as the timeout. So any number
+      // higher than the magical one will result in an unnecessarily slow test.
```

Comment:
I wasn't planning to add documentation until we're farther along. But I figured that `toDispatchEvents()` is unlikely to change much and felt that parts of it needed to be explained now.